zebra-chain/src/primitives/viewing_key.rs

```diff
@@ -0,0 +1,119 @@
+//! Type definitions for viewing keys and their hashes.
+
+use std::fmt;
+
+use zcash_primitives::{sapling::SaplingIvk, zip32::DiversifiableFullViewingKey};
+
+use crate::parameters::Network;
+
+pub mod sapling;
+
+#[cfg(test)]
+mod tests;
+
+const KEY_HASH_BYTE_SIZE: usize = 32;
+
+#[derive(Debug, Clone)]
+/// A Zcash Sapling or Orchard viewing key
+// TODO: Add Orchard types and any other Sapling key types
+pub enum ViewingKey {
+    /// An incoming viewing key for Sapling
+    SaplingIvk(Box<SaplingIvk>),
+
+    /// A diversifiable full viewing key for Sapling
+    DiversifiableFullViewingKey(Box<DiversifiableFullViewingKey>),
+}
+
+impl ViewingKey {
+    /// Returns an encoded byte representation of the viewing key
+    pub fn to_bytes(&self) -> Vec<u8> {
+        match self {
+            Self::SaplingIvk(ivk) => ivk.to_repr().to_vec(),
+            Self::DiversifiableFullViewingKey(dfvk) => dfvk.to_bytes().to_vec(),
+        }
+    }
+
+    /// Parses an encoded viewing key and returns it as a [`ViewingKey`] type.
+    pub fn parse(
```

Comment:
The doc comment says that the fn parses a viewing key, but since the fn relies on https://github.com/zcashfoundation/zebra/blob/297576f78ce1135d6db17d6e2f8b03d387672ff7/zebra-chain/src/primitives/viewing_key/sapling.rs#L32, it won't parse https://github.com/zcashfoundation/zebra/blob/297576f78ce1135d6db17d6e2f8b03d387672ff7/zebra-chain/src/primitives/viewing_key.rs#L20-L21.

Reply:
I don't think the `scan_block()` function in `zcash_client_backend` currently supports incoming viewing keys yet, I'll remove this for now and replace it with fvk / vk variants.

Reply:
Added more parse methods in https://github.com/ZcashFoundation/zebra/pull/8198/commits/b1d9723844be8cc42628a72beea0d7575ab8925b, mostly still unimplemented.